File: templates/WORKFLOW_RUNTIME_TEMPLATE.ts
// Workflow Runtime Template — TypeScript
//
// Use this as a starting point inside a Target Project when a POM workflow,
// Dynamic Workflow, or loop/goal YAML FSM needs executable integration points.
// Define and validate the FSM with any YAML that follows the POM workflow
// schema; WORKFLOW_TEMPLATE.yaml is only a reference starting point. Then run
// pom:workflow:lint and implement the ports below with target infrastructure.
// POM does not own these services at runtime.

export type Json = Record<string, unknown>;

export interface WorkflowSnapshot {
  workflow: string;
  version: number;
  state: string;
  context: Json;
  inFlight?: WorkflowHandle[];
}

export interface WorkflowHandle {
  name: string;
  workflow?: string;
  externalId?: string;
  status?: "active" | "awaited" | "cancelled" | "detached" | "suspended";
}

export interface ExecutionPort {
  dispatch(event: string, context: Json): Promise<Json> | Json;
}

export interface PersistencePort {
  saveSnapshot(key: string, snapshot: WorkflowSnapshot): Promise<void> | void;
  loadSnapshot(key: string): Promise<WorkflowSnapshot | undefined> | WorkflowSnapshot | undefined;
}

export interface TimerPort {
  schedule(key: string, delay: string, event: string): Promise<void> | void;
  cancel(key: string): Promise<void> | void;
  now(): Date;
}

export interface RetryPolicy {
  canRetry(key: string, context: Json): boolean;
  recordAttempt(key: string, context: Json): Json;
  nextDelay(key: string, context: Json): string | undefined;
}

export interface ToolPort {
  callTool(name: string, input: Json): Promise<Json>;
}

export interface SideEffectPort {
  publish(event: string, payload: Json): Promise<void> | void;
  compensate(step: string, context: Json): Promise<void> | void;
}

export interface RuntimePorts {
  execution: ExecutionPort;
  persistence: PersistencePort;
  timer: TimerPort;
  retry: RetryPolicy;
  tools: ToolPort;
  sideEffects: SideEffectPort;
}

export class WorkflowRuntime {
  constructor(private readonly ports: RuntimePorts) {}

  async step(key: string, event: string, snapshot: WorkflowSnapshot): Promise<WorkflowSnapshot> {
    const context = await this.ports.execution.dispatch(event, snapshot.context);
    const next: WorkflowSnapshot = { ...snapshot, context };
    await this.ports.persistence.saveSnapshot(key, next);
    await this.ports.sideEffects.publish("workflow.step", { key, workflow: next.workflow, state: next.state, event });
    return next;
  }

  async scheduleTimeout(key: string, delay: string, event: string): Promise<void> {
    await this.ports.timer.schedule(key, delay, event);
  }

  async callTool(name: string, input: Json): Promise<Json> {
    return this.ports.tools.callTool(name, input);
  }
}

File: templates/WORKFLOW_RUNTIME_TEMPLATE.py
"""Workflow Runtime Template — Python.

Use this as a starting point inside a Target Project when a POM workflow,
Dynamic Workflow, or loop/goal YAML FSM needs executable integration points.
Define and validate the FSM with any YAML that follows the POM workflow
schema; WORKFLOW_TEMPLATE.yaml is only a reference starting point. Then run
pom:workflow:lint and implement these protocols with target infrastructure.
POM does not own these services at runtime.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Protocol

Json = dict[str, Any]


@dataclass
class WorkflowHandle:
    name: str
    workflow: str | None = None
    external_id: str | None = None
    status: str | None = None


@dataclass
class WorkflowSnapshot:
    workflow: str
    version: int
    state: str
    context: Json
    in_flight: list[WorkflowHandle] = field(default_factory=list)


class ExecutionPort(Protocol):
    def dispatch(self, event: str, context: Json) -> Json: ...


class PersistencePort(Protocol):
    def save_snapshot(self, key: str, snapshot: WorkflowSnapshot) -> None: ...
    def load_snapshot(self, key: str) -> WorkflowSnapshot | None: ...


class TimerPort(Protocol):
    def schedule(self, key: str, delay: str, event: str) -> None: ...
    def cancel(self, key: str) -> None: ...
    def now(self) -> datetime: ...


class RetryPolicy(Protocol):
    def can_retry(self, key: str, context: Json) -> bool: ...
    def record_attempt(self, key: str, context: Json) -> Json: ...
    def next_delay(self, key: str, context: Json) -> str | None: ...


class ToolPort(Protocol):
    def call_tool(self, name: str, input: Json) -> Json: ...


class SideEffectPort(Protocol):
    def publish(self, event: str, payload: Json) -> None: ...
    def compensate(self, step: str, context: Json) -> None: ...


@dataclass
class RuntimePorts:
    execution: ExecutionPort
    persistence: PersistencePort
    timer: TimerPort
    retry: RetryPolicy
    tools: ToolPort
    side_effects: SideEffectPort


class WorkflowRuntime:
    def __init__(self, ports: RuntimePorts):
        self.ports = ports

    def step(self, key: str, event: str, snapshot: WorkflowSnapshot) -> WorkflowSnapshot:
        context = self.ports.execution.dispatch(event, snapshot.context)
        next_snapshot = WorkflowSnapshot(
            workflow=snapshot.workflow,
            version=snapshot.version,
            state=snapshot.state,
            context=context,
            in_flight=snapshot.in_flight,
        )
        self.ports.persistence.save_snapshot(key, next_snapshot)
        self.ports.side_effects.publish(
            "workflow.step",
            {"key": key, "workflow": next_snapshot.workflow, "state": next_snapshot.state, "event": event},
        )
        return next_snapshot

    def schedule_timeout(self, key: str, delay: str, event: str) -> None:
        self.ports.timer.schedule(key, delay, event)

    def call_tool(self, name: str, input: Json) -> Json:
        return self.ports.tools.call_tool(name, input)
